### ISSUE TITLE
more than once a second  failed TLS Handshake errors in log of kube-apiserver

### ISSUE THREAD

**Issue by uvwild** [gardener/gardener#3629]
/area control-plane
  /area networking
  /area security

/kind/bug
/kind/impediment
/priority/normal
    
The logs of the loadbalancer for api.otc-seed.garden.internal.garden.<OURBASEDOMAIN>
is full of TLS handshake errors since Feb 22.
`kubectl -n shoot--garden--otc-seed  logs -f $(k -n shoot--garden--otc-seed get pods -o name | grep kube-apiserver)  --all-containers --since=12h
`
`....
I0225 17:00:12.644545       1 log.go:172] http: TLS handshake error from 10.244.7.0:30464: read tcp 10.244.5.228:443->10.244.7.0:30464: read: connection reset by peer
....
`
**What you expected to happen**:
No TLS handshake errors

Being able to identify the  objects belonging to the IPs would help...
Or how to check the certificates.
I cannot only find find the IPs in my cluster. I checked ControlPlane/Node/Pod CIDRs. and  the provider VM
which is hosting the external loadbalancer. 

I read that there is in issue with the loadbalancer in front of the kube api server breaking TLS connections because its not aware of it. But I am not sure how I can use this information to reduce the error rate  (more than 1 / sec)

**Environment**:
- Gardener version:
API version v1.10.2   Dashboard version 1.46.1
- Kubernetes version (use `kubectl version`):
- Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T13:41:02Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:51:04Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
OTC

**Comment by rfranzke** [comment]
/cc @mvladev @DockToFuture 

@uvwild Why do you think this is a Gardener bug? It rather sounds more related to the underlying infrastructure, i.e., the load balancers of OTC, doesn't it?

**Comment by uvwild** [comment]
> /cc @mvladev @DockToFuture
> 
> @uvwild Why do you think this is a Gardener bug? It rather sounds more related to the underlying infrastructure, i.e., the load balancers of OTC, doesn't it?

The affected shoot cluster has been created in gardener at 2021-01-11  
TLS  error only start on 22.2.  

I also checked other shoot clusters they  also have this issue.
NS=shoot--garden--otc-seed
NS-shoot--jitsi--jitsi
kubectl -n $NS logs  $(k -n $NS get pods -o name | grep kube-apiserver) --all-containers   --since=12h

NS=kube-system
Only  the kube-api-server in the  kube-system of the cluster created in  the provider does not have the issue.

(Which does not rule out that its how the loadbalancer interacts with the gardener installed loadbalancers ;)

However, I am using this for some months now and problems only showed up recently.

**Comment by rfranzke** [comment]
> The affected shoot cluster has been created in gardener at 2021-01-11
TLS error only start on 22.2.

What exactly did change on 22.2. in your setup?

**Comment by timuthy** [comment]
Can these TLS errors come from a health monitor of the load balancer?

**Comment by uvwild** [comment]
In many cases the health checks seems to be cause such issues.

Therefore I disabled the health checks in  OTC and checked the logs.
But it did not affect the TLS handshake errors. 
What other health checks are there to be adjusted?

I see the TLS errors from the kube-api-server to various IPs on gateways (x.x.x.1) of the connected subnets/namespaces. 

On my side the resulting effects are broken helm commands. Those seem to be larger than the involved input buffers and keep on breaking. Whereas the kubectl seems unaffected in normal operation.

**Comment by uvwild** [comment]
I found a possible cause for this behaviour.  Sticky Sessions Disabled in the cloud provider load balancer.
I believe that would sufficiently explain the issues.

I added a    ` sessionAffinity: ClientIP`  
to my nginx ingress configuration which causes the sticky session in the cloud loadbalancer.

I  added sessions Affinity  to the kube-apiserver service of an affected k8s cluster:
```
  sessionAffinity: ClientIP
  sessionAffinityConfig:
    clientIP:
      timeoutSeconds: 10800
```
But it did not propagate a sticky session to the provider load balancer. 
and  the TLS handshalke errors within the kube cluster carried on. :(
**about 2 x per second!**

Anyone has an idea how to this can be forced to propagate. Deleting the kube-apiserver within a running cluster seems like a bad idea.

**Comment by uvwild** [comment]
I did some more reseach.

**The issue with the TLS handshake error persist in the OTC gardener cluster!**
`2021-03-17T00:09:46.280659281Z I0317 00:09:46.280566       1 log.go:172] http: TLS handshake error from 10.240.12.0:16384: read tcp 10.241.129.140:443->10.240.12.0:16384: read: connection reset by peerI analysed the jitsi cluster in more detail:Affected connections:10.241.129.140:443->10.240.12.0  `10.241.129.140:443->10.241.128.110.241.129.140:443->10.241.130.110.241.129.140:443->10.241.132.110.241.129.140:443->10.241.133.1
SRC:
shoot--jitsi--jitsi    kube-apiserver-6f9b77dc7b-wb7h6   \
10.241.129.140   shoot--garden--otc-seed-cpu-worker-z1-77848-28wlrDST
10.240.12.0   node shoot--garden--otc-seed-cpu-worker-z1-77848-28wlr
10.241.128.1 pods subnet gateway on shoot--garden--otc-seed-cpu-worker-z1-77848-plsj7
10.241.130.1 pods subnet gateway on shoot--garden--otc-seed-cpu-worker-z1-77848-sr54w
10.241.132.1 pods subnet gateway on shoot--garden--otc-seed-cpu-worker-z1-77848-6rxfc
10.241.133.1 pods subnet gateway on shoot--garden--otc-seed-cpu-worker-z1-77848-tx22l

All active clusters have the same issues:
```
NS=XXXXX
kubectl -n $NS logs  $(kubectl -n $NS get pods -o name | grep kube-apiserver | head -1) --all-containers --timestamps
```

The following issues point in the right direction. They talk about health checks between the instances and authentication.
https://github.com/kubernetes/kubernetes/issues/70411
https://github.com/kubernetes/kubernetes/issues/43784

I found a few failed healthcheck in a loadbalancer. But disabling them did not affect the logs. Also my  application ingress nginx loadbalancer shows the same TCP handshake errors. I start to believe now that this is broken systematically in  every shoot.The connection between the otc-seed-cluster and the shoots has those issues.

The first issue I encountered was the large helm commands. (kube-apiserver)

**But it seems to break high volume user k8s applications as well**
Is there any k8s network specialist around? Can we setup some meeting? 

In my k8s app it is the same pattern on the loadbalancer.
Communication from  the  **node ip of the load balancer** and the **podnet gateways** to the loadbalancer is impaired.
Tthe nginx log show ssl handshake errors on 443 and connection while waiting for request for port 80 connections:

```
[info] 38#38: *120326 recv() failed (104: Connection reset by peer) while SSL handshaking, client: 100.96.2.1, server: 0.0.0.0:443
[info] 39#39: *120337 recv() failed (104: Connection reset by peer) while waiting for request, client: 100.96.2.1, server: 0.0.0.0:80
```
I dont know what I can do about it since I dont know the source of the failed connections or how to analyze them further.
**The pod subnet gateway IP involved is not a pode I can access.**
